File: packages/ruby-sdk/lib/strait/operations/sdk_runs.rb
# frozen_string_literal: true

module Strait
  module Operations
    # Service for SDK-side run lifecycle operations (used by worker SDKs).
    class SDKRunsService < BaseService
      # Annotate a run with metadata.
      # POST /sdk/v1/runs/{runID}/annotate
      def annotate_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/annotate", path_params: { "runID" => run_id }, body: body)
      end

      # Create a checkpoint for a run.
      # POST /sdk/v1/runs/{runID}/checkpoint
      def checkpoint_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/checkpoint", path_params: { "runID" => run_id }, body: body)
      end

      # Mark a run as complete.
      # POST /sdk/v1/runs/{runID}/complete
      def complete_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/complete", path_params: { "runID" => run_id }, body: body)
      end

      # Continue a suspended run.
      # POST /sdk/v1/runs/{runID}/continue
      def continue_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/continue", path_params: { "runID" => run_id }, body: body)
      end

      # Mark a run as failed.
      # POST /sdk/v1/runs/{runID}/fail
      def fail_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/fail", path_params: { "runID" => run_id }, body: body)
      end

      # Send a heartbeat for a run.
      # POST /sdk/v1/runs/{runID}/heartbeat
      def heartbeat_run(run_id)
        _request(:post, "/sdk/v1/runs/{runID}/heartbeat", path_params: { "runID" => run_id })
      end

      # Append a log entry to a run.
      # POST /sdk/v1/runs/{runID}/log
      def log_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/log", path_params: { "runID" => run_id }, body: body)
      end

      # Record an output for a run.
      # POST /sdk/v1/runs/{runID}/output
      def output_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/output", path_params: { "runID" => run_id }, body: body)
      end

      # Report progress for a run.
      # POST /sdk/v1/runs/{runID}/progress
      def progress_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/progress", path_params: { "runID" => run_id }, body: body)
      end

      # Spawn a child run.
      # POST /sdk/v1/runs/{runID}/spawn
      def spawn_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/spawn", path_params: { "runID" => run_id }, body: body)
      end

      # Record a tool call for a run.
      # POST /sdk/v1/runs/{runID}/tool-call
      def tool_call_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/tool-call", path_params: { "runID" => run_id }, body: body)
      end

      # Report resource usage for a run.
      # POST /sdk/v1/runs/{runID}/usage
      def usage_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/usage", path_params: { "runID" => run_id }, body: body)
      end

      # Wait for an event during a run.
      # POST /sdk/v1/runs/{runID}/wait-for-event
      def wait_for_event_run(run_id, body)
        _request(:post, "/sdk/v1/runs/{runID}/wait-for-event", path_params: { "runID" => run_id }, body: body)
      end
    end
  end
end

File: packages/rust-sdk/src/operations/sdk_runs.rs
use crate::client::StraitClient;
use crate::errors::StraitError;
use crate::http::substitute_path_params;
use std::sync::Arc;

pub struct SdkRunsService {
    client: Arc<StraitClient>,
}

impl SdkRunsService {
    pub fn new(client: Arc<StraitClient>) -> Self {
        Self { client }
    }

    pub async fn annotate_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/annotate", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn checkpoint_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/checkpoint", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn complete_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/complete", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn continue_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/continue", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn fail_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/fail", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn heartbeat_run(&self, run_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/heartbeat", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, None)
            .await
    }

    pub async fn log_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/log", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn output_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/output", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn progress_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/progress", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn spawn_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/spawn", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn tool_call_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/tool-call", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn usage_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/usage", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn wait_for_event_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path =
            substitute_path_params("/sdk/v1/runs/{runID}/wait-for-event", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn set_state(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/state", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn list_state(&self, run_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/state", &[("runID", run_id)]);
        self.client
            .do_request("GET", &path, None, None, None)
            .await
    }

    pub async fn get_state(
        &self,
        run_id: &str,
        key: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/sdk/v1/runs/{runID}/state/{key}",
            &[("runID", run_id), ("key", key)],
        );
        self.client
            .do_request("GET", &path, None, None, None)
            .await
    }

    pub async fn delete_state(
        &self,
        run_id: &str,
        key: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/sdk/v1/runs/{runID}/state/{key}",
            &[("runID", run_id), ("key", key)],
        );
        self.client
            .do_request("DELETE", &path, None, None, None)
            .await
    }

    pub async fn get_payload(&self, run_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/payload", &[("runID", run_id)]);
        self.client
            .do_request("GET", &path, None, None, None)
            .await
    }

    pub async fn resources_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/resources", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn stream_run(
        &self,
        run_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/sdk/v1/runs/{runID}/stream", &[("runID", run_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }
}
